src/mnemonic/mod.rs

```diff
@@ -35,28 +53,69 @@ pub fn mnemonic_to_entropy(words: Vec<String>) -> Result<[u8; 32]> {
 
     let divider_index: usize = ((bits.len() as f64 / 33.0) * 32.0).floor() as usize;
     let (entropy_bits, checksum_bits) = bits.split_at(divider_index);
-    // The mobile wallet does not calculate the checksum bits right so
-    // they always and up being all 0
-    if checksum_bits != "0000" {
-        bail!("invalid checksum");
-    }
 
     lazy_static! {
         static ref RE_BYTES: Regex = Regex::new("(.{1,8})").unwrap();
     }
 
-    let mut entropy_base = [0u8; 16];
-    for (idx, matched) in RE_BYTES.find_iter(&entropy_bits).enumerate() {
-        entropy_base[idx] = binary_to_bytes(matched.as_str()) as u8;
+    // For up to 24 words, checksum should only ever be a single byte
+    let mut checksum_bytes = [0u8; 1];
+    for (idx, matched) in RE_BYTES.find_iter(&checksum_bits).enumerate() {
+        checksum_bytes[idx] = binary_to_bytes(matched.as_str()) as u8;
     }
 
     let mut entropy_bytes = [0u8; 32];
-    entropy_bytes[..16].copy_from_slice(&entropy_base);
-    entropy_bytes[16..].copy_from_slice(&entropy_base);
+    let valid_checksum;
+    if words.len() == 12 {
+        let mut entropy_base = [0u8; 16];
+        for (idx, matched) in RE_BYTES.find_iter(&entropy_bits).enumerate() {
+            entropy_base[idx] = binary_to_bytes(matched.as_str()) as u8;
+        }
+
+        // If this is supposed to be a BIP39 12-word phrase, verify the
+        // checksum. Otherwise assume it is a phrase from the mobile app.
+        // The mobile wallet does not calculate the checksum bits right so
+        // its checksums are always 0
+        valid_checksum = match seed_type {
+            SeedType::Bip39 => calc_checksum_128(entropy_base),
+            SeedType::Mobile => 0,
+        };
+        entropy_bytes[..16].copy_from_slice(&entropy_base);
+        entropy_bytes[16..].copy_from_slice(&entropy_base);
+    } else {
+        for (idx, matched) in RE_BYTES.find_iter(&entropy_bits).enumerate() {
+            entropy_bytes[idx] = binary_to_bytes(matched.as_str()) as u8;
+        }
+
+        // 24-word phrases can't be from the mobile wallet so it should have
+        // a valid BIP39 checksum.
+        valid_checksum = calc_checksum_256(entropy_bytes);
+    }
+
+    if checksum_bytes[0] != valid_checksum {
+        bail!("Checksum failed. Invalid seed phrase.");
+    }
 
     Ok(entropy_bytes)
 }
 
+fn calc_checksum_128(bytes: [u8; 16]) -> u8 {
+    let mut hash = Sha256::new();
```

Comment:
You already imported sha2::Digest, so use it? `Sha256::digest(bytes)` will make this one line

Reply:
nice catch, thanks! I fixed this, but left these as functions rather than just using `Sha256::digest()` at the call-sites because I think it makes it easier to understand what is going on and maybe easier to add another checksum algo if there ends up being another SeedType.